src/secp256k1.c

```diff
@@ -446,11 +446,91 @@ static int nonce_function_rfc6979(unsigned char *nonce32, const unsigned char *m
 
+int secp256k1_ecdsa_anti_nonce_sidechan_client_commit(
+    const secp256k1_context* ctx,
+    secp256k1_pubkey *client_commit,
+    const unsigned char *msg32,
+    const unsigned char *seckey32,
+    unsigned char *rand_commitment32
+) {
+    unsigned char nonce32[32];
+    secp256k1_scalar k;
+    secp256k1_gej rj;
+    secp256k1_ge r;
+    unsigned int count = 0;
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(client_commit != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(seckey32 != NULL);
+    ARG_CHECK(rand_commitment32 != NULL);
+
+
+    while (1) {
+        int overflow = 0;
+        if (!secp256k1_nonce_function_default(nonce32, msg32, seckey32, NULL, rand_commitment32, count)) {
+            return 0;
+        }
+
+        secp256k1_scalar_set_b32(&k, nonce32, &overflow);
+        if (!overflow && !secp256k1_scalar_is_zero(&k)) {
+            break;
+        }
+        count++;
+    }
+
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &rj, &k);
+    secp256k1_ge_set_gej(&r, &rj);
+    secp256k1_pubkey_save(client_commit, &r);
+    return 1;
+}
+
 int secp256k1_ecdsa_sign(const secp256k1_context* ctx, secp256k1_ecdsa_signature *signature, const unsigned char *msg32, const unsigned char *seckey, secp256k1_nonce_function noncefp, const void* noncedata) {
+    return secp256k1_ecdsa_sign_to_contract(ctx, signature, msg32, seckey, noncefp, noncedata, NULL);
+}
+
+/* Compute an ec commitment tweak as hash(pubkey, data). */
+static int secp256k1_ec_commit_tweak(unsigned char *tweak32, const secp256k1_ge *pubkey, const unsigned char *data, size_t data_size) {
+    secp256k1_ge p = *pubkey;
+    unsigned char rbuf[33];
+    size_t rbuf_size = sizeof(rbuf);
+    secp256k1_sha256 sha;
+
+    if (data_size == 0) {
+        /* That's probably not what the caller wanted */
+        return 0;
+    }
+    secp256k1_eckey_pubkey_serialize(&p, rbuf, &rbuf_size, 1);
+
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_sha256_write(&sha, rbuf, rbuf_size);
+    secp256k1_sha256_write(&sha, data, data_size);
+    secp256k1_sha256_finalize(&sha, tweak32);
+    return 1;
+}
+
+/* Compute an ec commitment tweak as hash(pubkey, data). */
+static int secp256k1_ec_commit_tweak_from_seckey(
+    const secp256k1_context *ctx,
+    unsigned char *tweak32,
+    const secp256k1_scalar *seckey,
+    const unsigned char *data,
+    size_t data_size) {
+    secp256k1_gej rp;
+    secp256k1_ge r;
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &rp, seckey);
+    secp256k1_ge_set_gej(&r, &rp);
+    return secp256k1_ec_commit_tweak(tweak32, &r, data, data_size);
+}
+
+int secp256k1_ecdsa_sign_to_contract(const secp256k1_context* ctx, secp256k1_ecdsa_signature *signature, const unsigned char *msg32, const unsigned char *seckey, secp256k1_nonce_function noncefp, const void* noncedata, const unsigned char* s2c_data32) {
```

Comment:
So right now you can't get the original nonce when using pure sign-to-contract, unless you do `secp256k1_ecdsa_anti_nonce_sidechan_client_commit` ?

Reply:
Yeah, Part of it is that it seemed redundant, and part of it is that I am not very knowledgeable about sign-to-contract, e.g. what other uses of it there are where one would not use something akin to `secp256k1_ecdsa_anti_nonce_sidechan_client_commit` beforehand anyway.

I can add an opening struct here too if you think it will be helpful, let me know.

Reply:
I don't know a specific application for s2c besides the anti nonce sidechannel and committing to the signer set in threshold signatures. I'd add an opening struct argument.